Regression in jumpengine

**Describe the bug**
Slice index out of range:

```
panic: runtime error: index out of range [4] with length 4

goroutine 1 [running]:
github.com/retroenv/nesgodisasm/internal/jumpengine.(*JumpEngine).ScanForNewJumpEngineEntry(0x400010e480, {0x199ea0, 0x4000000360})
        /home/chris/git/nespp/THIRD_PARTY/disassembler/internal/jumpengine/jumpengine.go:249 +0x5e4
github.com/retroenv/nesgodisasm/internal.(*Disasm).addressToDisassemble(0x4000000360)
        /home/chris/git/nespp/THIRD_PARTY/disassembler/internal/parser.go:93 +0x50
github.com/retroenv/nesgodisasm/internal.(*Disasm).followExecutionFlow(0x4000000360)
        /home/chris/git/nespp/THIRD_PARTY/disassembler/internal/parser.go:14 +0x28
github.com/retroenv/nesgodisasm/internal.(*Disasm).Process(0x4000000360, {0x197f38, 0x40000540c8}, 0x4000011aa0)
        /home/chris/git/nespp/THIRD_PARTY/disassembler/internal/disasm.go:103 +0x30
main.processFile(0x40000115f0, {{0x15f717, 0x4}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0xffffed4ee885, ...}, ...}, ...)
        /home/chris/git/nespp/THIRD_PARTY/disassembler/main.go:235 +0x17c
main.disasmFile(0x40000115f0, {{0x15f717, 0x4}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0xffffed4ee885, ...}, ...}, ...)
        /home/chris/git/nespp/THIRD_PARTY/disassembler/main.go:214 +0x4dc
main.main()
        /home/chris/git/nespp/THIRD_PARTY/disassembler/main.go:52 +0x258
```

**To Reproduce**
Steps to reproduce the behavior (used ROM filename/CRC32 and command-line parameters):

- I used http://nickmass.com/images/nestest.nes, but also repro'd with Kirby's Adventure (sha256: `b0d74f4270ad5855f10224c460956f6dfe41d6673a38262ad0d37aa9ce103dd5`)
- CLI invocation `go build . && ./nesgodisasm -q ../../rom.nes`

**Expected behavior**
No panic

**System (please complete the following information):**
 - OS: linux arm64 ubuntu
 - Version / Commit: I bisected the regression to: https://github.com/retroenv/nesgodisasm/commit/23f547cabb64bafabcb7d11e56964ecd89f21b3f

**Additional context**
I believe the bug is:

```diff
-		for i := 0; i < len(j.jumpEngineCallers); i++ {
+		for i := range j.jumpEngineCallers {
			engineCaller := j.jumpEngineCallers[i]
			if engineCaller.terminated {
				// jump engine table is processed, remove it from list to process
-				j.jumpEngineCallers = append(j.jumpEngineCallers[:i], j.jumpEngineCallers[i+1:]...)
+				j.jumpEngineCallers = slices.Delete(j.jumpEngineCallers, i, i+1)
			}
```

I believe it's a mistake to mutate a slice while iterating over it in a range call.

I'm guessing the original for loop worked because we keep re-evaluate the slice length on each iteration, although I feel like we might be skipping a caller when we re-assign the slice, since the the previous `i` now points to what should be the next caller.

I think the right thing to do here is to make a copy of the slice that we iterate over, and delete from the original. I'll open a PR.

--- a/internal/arch/mocks/disasm.go
+++ b/internal/arch/mocks/disasm.go
@@ -1,0 +1,101 @@
+// Package mocks provides mock implementations of arch interfaces for testing.
+package mocks
+
+import (
+	"github.com/retroenv/nesgodisasm/internal/arch"
+	"github.com/retroenv/nesgodisasm/internal/options"
+	"github.com/retroenv/nesgodisasm/internal/program"
+	"github.com/retroenv/retrogolib/arch/system/nes/cartridge"
+	"github.com/retroenv/retrogolib/log"
+)
+
+// Disasm is a minimal mock implementation of arch.Disasm for testing.
+type Disasm struct {
+	logger   *log.Logger
+	mapper   arch.Mapper
+	codeBase uint16
+	Memory   []byte
+}
+
+// NewDisasm creates a new mock Disasm with the given logger and code base address.
+func NewDisasm(logger *log.Logger, mapper arch.Mapper, codeBase uint16, memorySize int) *Disasm {
+	return &Disasm{
+		logger:   logger,
+		mapper:   mapper,
+		codeBase: codeBase,
+		Memory:   make([]byte, memorySize),
+	}
+}
+
+func (m *Disasm) AddAddressToParse(address, context, from uint16, currentInstruction arch.Instruction, isABranchDestination bool) {
+}
+
+func (m *Disasm) Cart() *cartridge.Cartridge {
+	return nil
+}
+
+func (m *Disasm) ChangeAddressRangeToCodeAsData(address uint16, data []byte) {
+}
+
+func (m *Disasm) CodeBaseAddress() uint16 {
+	return m.codeBase
+}
+
+func (m *Disasm) Constants() arch.ConstantManager {
+	return nil
+}
+
+func (m *Disasm) DeleteFunctionReturnToParse(uint16) {
+}
+
+func (m *Disasm) JumpEngine() arch.JumpEngine {
+	return nil
+}
+
+func (m *Disasm) Logger() *log.Logger {
+	return m.logger
+}
+
+func (m *Disasm) Mapper() arch.Mapper {
+	return m.mapper
+}
+
+func (m *Disasm) Options() options.Disassembler {
+	return options.Disassembler{}
+}
+
+func (m *Disasm) ProgramCounter() uint16 {
+	return 0
+}
+
+func (m *Disasm) ReadMemory(address uint16) (byte, error) {
+	if int(address) >= len(m.Memory) {
+		return 0, nil
+	}
+	return m.Memory[address], nil
+}
+
+func (m *Disasm) ReadMemoryWord(address uint16) (uint16, error) {
+	low, err := m.ReadMemory(address)
+	if err != nil {
+		return 0, err
+	}
+	high, err := m.ReadMemory(address + 1)
+	if err != nil {
+		return 0, err
+	}
+	return uint16(low) | uint16(high)<<8, nil
+}
+
+func (m *Disasm) SetCodeBaseAddress(uint16) {
+}
+
+func (m *Disasm) SetHandlers(program.Handlers) {
+}
+
+func (m *Disasm) SetVectorsStartAddress(uint16) {
+}
+
+func (m *Disasm) Variables() arch.VariableManager {
+	return nil
+}

--- a/internal/arch/mocks/mapper.go
+++ b/internal/arch/mocks/mapper.go
@@ -1,0 +1,49 @@
+package mocks
+
+import (
+	"github.com/retroenv/nesgodisasm/internal/arch"
+	"github.com/retroenv/nesgodisasm/internal/program"
+)
+
+// Mapper is a minimal mock implementation of arch.Mapper for testing.
+type Mapper struct {
+	offsets map[uint16]*arch.Offset
+}
+
+// NewMapper creates a new mock Mapper.
+func NewMapper() *Mapper {
+	return &Mapper{
+		offsets: make(map[uint16]*arch.Offset),
+	}
+}
+
+func (m *Mapper) GetMappedBank(uint16) arch.MappedBank {
+	return nil
+}
+
+func (m *Mapper) GetMappedBankIndex(uint16) uint16 {
+	return 0
+}
+
+func (m *Mapper) ReadMemory(address uint16) byte {
+	return 0
+}
+
+func (m *Mapper) OffsetInfo(address uint16) *arch.Offset {
+	if offset, ok := m.offsets[address]; ok {
+		return offset
+	}
+	offset := &arch.Offset{
+		Offset: program.Offset{
+			Type: program.DataOffset,
+			Data: []byte{0},
+		},
+	}
+	m.offsets[address] = offset
+	return offset
+}
+
+// SetOffsetInfo sets the offset info for the given address (for testing purposes).
+func (m *Mapper) SetOffsetInfo(address uint16, offset *arch.Offset) {
+	m.offsets[address] = offset
+}

--- a/internal/jumpengine/jumpengine.go
+++ b/internal/jumpengine/jumpengine.go
@@ -241,18 +241,16 @@ func (j *JumpEngine) ScanForNewJumpEngineEntry(dis arch.Disasm) (bool, error) {
 	logger := dis.Logger()
 
 	for len(j.jumpEngineCallers) != 0 {
+		// Remove all terminated entries
+		j.jumpEngineCallers = slices.DeleteFunc(j.jumpEngineCallers, func(ec *jumpEngineCaller) bool {
+			return ec.terminated
+		})
+
+		// Find the jump engine table with the smallest number of processed entries.
+		// This conservative approach avoids interpreting code in the table area as function references.
 		minEntries := -1
-
-		// find the jump engine table with the smallest number of processed entries,
-		// this conservative approach avoids interpreting code in the table area as function references
-		for i := range j.jumpEngineCallers {
-			engineCaller := j.jumpEngineCallers[i]
-			if engineCaller.terminated {
-				// jump engine table is processed, remove it from list to process
-				j.jumpEngineCallers = slices.Delete(j.jumpEngineCallers, i, i+1)
-			}
-
-			if i := engineCaller.entries; !engineCaller.terminated && (i < minEntries || minEntries == -1) {
+		for _, engineCaller := range j.jumpEngineCallers {
+			if i := engineCaller.entries; i < minEntries || minEntries == -1 {
 				minEntries = i
 			}
 		}

--- a/internal/jumpengine/jumpengine_test.go
+++ b/internal/jumpengine/jumpengine_test.go
@@ -1,0 +1,91 @@
+package jumpengine
+
+import (
+	"testing"
+
+	"github.com/retroenv/nesgodisasm/internal/arch/m6502"
+	"github.com/retroenv/nesgodisasm/internal/arch/mocks"
+	"github.com/retroenv/nesgodisasm/internal/assembler/ca65"
+	"github.com/retroenv/retrogolib/arch/system/nes/parameter"
+	"github.com/retroenv/retrogolib/assert"
+	"github.com/retroenv/retrogolib/log"
+)
+
+// TestScanForNewJumpEngineEntry_MultipleTerminated verifies that multiple terminated
+// jump engine callers are correctly removed from the processing queue.
+func TestScanForNewJumpEngineEntry_MultipleTerminated(t *testing.T) {
+	logger := log.NewTestLogger(t)
+	ar := m6502.New(parameter.New(ca65.ParamConfig))
+	mapper := mocks.NewMapper()
+	dis := mocks.NewDisasm(logger, mapper, 0x8000, 0x10000)
+	je := New(ar)
+
+	// Create multiple terminated jump engine callers that should all be removed.
+	caller1 := &jumpEngineCaller{
+		tableStartAddress: 0x8010,
+		entries:           2,
+		terminated:        true,
+	}
+	caller2 := &jumpEngineCaller{
+		tableStartAddress: 0x8020,
+		entries:           3,
+		terminated:        true,
+	}
+	caller3 := &jumpEngineCaller{
+		tableStartAddress: 0x8030,
+		entries:           1,
+		terminated:        true,
+	}
+	je.jumpEngineCallers = []*jumpEngineCaller{caller1, caller2, caller3}
+
+	found, err := je.ScanForNewJumpEngineEntry(dis)
+
+	assert.NoError(t, err)
+	assert.False(t, found, "should not find any new entries when all are terminated")
+	assert.Len(t, je.jumpEngineCallers, 0, "all terminated entries should be removed")
+}
+
+// TestScanForNewJumpEngineEntry_MixedTerminated tests the scenario where
+// some entries are terminated and others are not.
+func TestScanForNewJumpEngineEntry_MixedTerminated(t *testing.T) {
+	logger := log.NewTestLogger(t)
+	ar := m6502.New(parameter.New(ca65.ParamConfig))
+	mapper := mocks.NewMapper()
+	dis := mocks.NewDisasm(logger, mapper, 0x8000, 0x10000)
+
+	// Set up memory with a valid function reference
+	dis.Memory[0x8030] = 0x00 // Low byte
+	dis.Memory[0x8031] = 0x80 // High byte (points to 0x8000, within code range)
+
+	je := New(ar)
+
+	// Mix of terminated and active entries
+	caller1 := &jumpEngineCaller{
+		tableStartAddress: 0x8010,
+		entries:           2,
+		terminated:        true,
+	}
+	caller2 := &jumpEngineCaller{
+		tableStartAddress: 0x8020,
+		entries:           0,
+		terminated:        false,
+	}
+	caller3 := &jumpEngineCaller{
+		tableStartAddress: 0x8030,
+		entries:           0,
+		terminated:        false,
+	}
+	je.jumpEngineCallers = []*jumpEngineCaller{caller1, caller2, caller3}
+
+	found, err := je.ScanForNewJumpEngineEntry(dis)
+
+	assert.NoError(t, err)
+	// The active entries should remain
+	assert.Len(t, je.jumpEngineCallers, 1)
+	assert.Equal(t, je.jumpEngineCallers[0], caller3)
+	// At least one of the active entries should be processed
+	assert.True(t, found)
+	assert.Equal(t, 2, caller1.entries)
+	assert.Equal(t, 0, caller2.entries)
+	assert.Equal(t, 1, caller3.entries)
+}

